### PR TITLE
Upvote downvote

### PR DESCRIPTION
**Pull request by Freitag-fri**

### CODE REVIEW

File: contracts/libraries/VoteLib.sol [hunk 1]
```diff
@@ -0,0 +1,144 @@
+pragma solidity >=0.5.0;
+pragma abicoder v2;
+
+import "./PostLib.sol";
+
+/// @title VoteLib
+/// @notice Provides information about operation with posts                     //
+/// @dev posts information is stored in the mapping on the main contract        ///
+library VoteLib  {
+    enum VoteResource { Downvote, Upvoted, Downvoted, CorrecReply, FirstReply, Reply15Minutes }
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: Needs better name. `VoteAction`, `ResourceAction` or simply `Action`?

Comment (sergeyi23) [reply contracts/libraries/VoteLib.sol]: Rename `CorrecReply` to `AcceptReply`

File: contracts/libraries/VoteLib.sol [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    int8 constant UpvotedTutorial = 0;
+    int8 constant DownvotedTutorial = 0;
+
+/////////////////////////////////////////////////////////////////////////////
+
+    //expert reply
+    int8 constant DownvoteExpertReply = 0;
+    int8 constant UpvotedExpertReply = 0;
+    int8 constant DownvotedExpertReply = 0;
+    int8 constant CorrectExpertReply = 0;
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: `AcceptExpertReply`

File: contracts/libraries/VoteLib.sol (hunk 1 above)
Comment (sergeyi23) [reply contracts/libraries/VoteLib.sol]: or `BestReply`

File: contracts/libraries/VoteLib.sol [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            else if (VoteResource.Upvoted == voteResource) return UpvotedCommonPost;
+            else if (VoteResource.Downvoted == voteResource) return DownvotedCommonPost;
+
+        } else if (PostLib.TypePost.Tutorial == typePost) {
+            if (VoteResource.Downvote == voteResource) return DownvoteTutorial;
+            else if (VoteResource.Upvoted == voteResource) return UpvotedTutorial;
+            else if (VoteResource.Downvoted == voteResource) return DownvotedTutorial;
+
+        }
+        require(false, "TypePost or voteResource dont found");
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: TypePost or voteResource is not found

File: contracts/libraries/VoteLib.sol [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        } else if (PostLib.TypePost.CommonPost == typePost) {
+            if (VoteResource.Downvote == voteResource) return DownvoteCommonReply;
+            else if (VoteResource.Upvoted == voteResource) return UpvotedCommonReply;
+            else if (VoteResource.Downvoted == voteResource) return DownvotedCommonReply;
+            else if (VoteResource.CorrecReply == voteResource) return CorrectCommonReply;
+            else if (VoteResource.FirstReply == voteResource) return FirstCommonReply;
+            else if (VoteResource.Reply15Minutes == voteResource) return Reply15MinutesCommon;
+
+        }
+        return 0;
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: No require here?

File: contracts/libraries/VoteLib.sol [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    /// @param user user who voted for content
+    /// @param historyVote history vote all users
+    function getHistoryVote(
+        address user,
+        mapping(address => int256) storage historyVote
+    ) private view returns (int256) {
+        return historyVote[user];
+    }
+
+    function changeHistory(
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: Rename to `getForumItemRatingChange`

File: contracts/libraries/VoteLib.sol [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    int8 constant UpvotedCommonReply = 0;
+    int8 constant DownvotedCommonReply = 0;
+    int8 constant CorrectCommonReply = 0;
+    int8 constant FirstCommonReply = 0;
+    int8 constant Reply15MinutesCommon = 0;
+
+    /// @notice Get value Rating for post
+    /// @param typePost Type post: expertPost, commonPost, tutorial
+    /// @param voteResource Rating action: Downvote, Upvoted, Downvoted
+    function getRatingPost(
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: `getUserRatingChangeForPostAction`

File: contracts/libraries/VoteLib.sol [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            else if (VoteResource.Downvoted == voteResource) return DownvotedTutorial;
+
+        }
+        require(false, "TypePost or voteResource dont found");      
+    }
+
+    /// @notice Get value Rating for rating
+    /// @param typePost Type post: expertPost, commonPost, tutorial
+    /// @param voteResource Rating action: Downvote, Upvoted, Downvoted, CorrecReply...
+    function getRatingReply(
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol]: getUserRatingChangeForReplyAction

File: contracts/libraries/PostLib.sol [hunk 8]
```diff
@@ -329,6 +342,93 @@ library PostLib  {
...
+        post.info.rating += changeRating;
+    }
+ 
+    function voteReply(
+        UserLib.UserCollection storage users,
+        ReplyContainer storage reply,
+        address votedUser,
+        TypePost typePost,
+        bool isUpvote
+    ) internal {
```
Comment (sergeyi23) [inline contracts/libraries/PostLib.sol]: Why `votePost` is private and this is internal?

File: contracts/libraries/VoteLib.sol [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        mapping(address => int256) storage historyVote
+    ) private view returns (int256) {
+        return historyVote[user];
+    }
+
+    function changeHistory(
+        address actionAddress,
+        mapping(address => int256) storage historyVotes,
+        bool isUpvote
+    ) internal returns (int8) {
```
Comment (sergeyi23) [inline contracts/libraries/VoteLib.sol:117]: `view` function

**Review by sergeyi23 — CHANGES_REQUESTED** [review]
See the comments.

Also missing events for all of the actions.

File: contracts/libraries/VoteLib.sol (hunk 9 above)
Comment (Freitag-fri) [reply contracts/libraries/VoteLib.sol:117]: in this function changed value historyVotes

File: contracts/libraries/PostLib.sol [hunk 10]
```diff
@@ -64,13 +74,26 @@ library PostLib  {
...
 
     struct PostCollection {
         mapping(uint32 => PostContainer) posts;    // uint32?
         uint32 postCount;
     }
 
+    event PostCreated(address user, uint8 communityId, uint256 postId, bytes32 ipfsHash);
+    event ReplyCreated(address user, uint256 postId, uint16[] path, uint256 replyId, bytes32 ipfsHash);
+    event CommentCreated(address user, uint256 postId, uint16[] path, uint256 commentId, bytes32 ipfsHash);
+    event PostEditeded(address user, uint256 communityId, uint256 postId, bytes32 ipfsHash);
```
Comment (sergeyi23) [inline contracts/libraries/PostLib.sol]: Rename to `PostEdited`

File: contracts/libraries/PostLib.sol [hunk 11]
```diff
@@ -64,13 +74,26 @@ library PostLib  {
...
+    event ReplyEditeded(address user, uint256 postId, uint16[] path, uint256 replyId, bytes32 ipfsHash);
```
Comment (sergeyi23) [inline contracts/libraries/PostLib.sol]: Rename to `ReplyEdited`

File: contracts/libraries/PostLib.sol [hunk 12]
```diff
@@ -64,13 +74,26 @@ library PostLib  {
...
+    event CommentEdit(address user, uint256 postId, uint16[] path, uint256 commentId, bytes32 ipfsHash);
```
Comment (sergeyi23) [inline contracts/libraries/PostLib.sol]: Rename to `CommentEdited`

**Review by sergeyi23 — APPROVED** [review]
Fix event names before merging

**Pull request merged**